File: p977e/sorted_squares.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <vector>
#include <iostream>

using namespace std;

class Solution
{
public:
    vector<int> sortedSquares(vector<int> &nums)
    {
        vector<int> ans(nums.size(), 0);
        for (int l = 0, r = nums.size() - 1, i = r; i >= 0; i--)
        {
            int lsq = nums[l] * nums[l], rsq = nums[r] * nums[r];
            if (lsq > rsq)
            {
                ans[i] = lsq;
                l++;
            }
            else
            {
                ans[i] = rsq;
                r--;
            }
        }
        return ans;
    }
};

int main(int argc, char *argv[])
{
    vector<vector<int>> tests = {
        {-4, -1, 0, 3, 10},
        {-7, -3, 2, 3, 11}};
    Solution sol;
    for (auto &t : tests)
    {
        vector<int> actual = sol.sortedSquares(t);
        cout << "Squares of a sorted array -> [";
        for (int n : actual)
            cout << n << ",";
        cout << "]" << endl;
    }
    return 0;
}

File: p977e/sorted_squares.py
from typing import List


class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        sq, l, r = [], 0, len(nums) - 1
        while l <= r:
            if abs(nums[l]) <= abs(nums[r]):
                sq.append(nums[r] * nums[r])
                r -= 1
            else:
                sq.append(nums[l] * nums[l])
                l += 1
        return sq[::-1]


# TESTS
for nums, expected in [
    ([-4, -1, 0, 3, 10], [0, 1, 9, 16, 100]),
    ([-7, -3, 2, 3, 11], [4, 9, 9, 49, 121]),
]:
    sol = Solution()
    actual = sol.sortedSquares(nums)
    print("Sorted squares of", nums, "->", actual)
    assert actual == expected
